File: spec/factories.rb
# frozen_string_literal: true

#
# Copyright (C) 2020 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

#
# To use factories in a development rails console, run
#   require 'spec/factories'
#   include Factories
#

module Factories
  def factory_with_protected_attributes(ar_klass, attrs, do_save = true)
    obj = ar_klass.respond_to?(:new) ? ar_klass.new : ar_klass.build
    attrs.each { |k, v| obj.send(:"#{k}=", v) }
    obj.save! if do_save
    obj
  end

  def update_with_protected_attributes!(ar_instance, attrs)
    attrs.each { |k, v| ar_instance.send(:"#{k}=", v) }
    ar_instance.save!
  end

  def update_with_protected_attributes(ar_instance, attrs)
    update_with_protected_attributes!(ar_instance, attrs) rescue false
  end

  # a fast way to create a record, especially if you don't need the actual
  # ruby object. since it just does a straight up insert, you need to
  # provide any non-null attributes or things that would normally be
  # inferred/defaulted prior to saving
  def create_record(klass, attributes, return_type = :id)
    create_records(klass, [attributes], return_type)[0]
  end

  # a little wrapper around bulk_insert that gives you back records or ids
  # in order
  # NOTE: if you decide you want to go add something like this to canvas
  # proper, make sure you have it handle concurrent inserts (this does
  # not, because READ COMMITTED is the default transaction isolation
  # level)
  def create_records(klass, records, return_type = :id)
    return [] if records.empty?

    klass.transaction do
      klass.connection.bulk_insert klass.table_name, records
      next if return_type == :nil

      scope = klass.order("id DESC").limit(records.size)
      if return_type == :record
        scope.to_a.reverse
      else
        scope.pluck(:id).reverse
      end
    end
  end
end

legit_global_methods = Object.private_methods
Dir[File.dirname(__FILE__) + "/factories/**/*.rb"].each { |f| require f }
crap_factories = (Object.private_methods - legit_global_methods)
if crap_factories.present?
  warn "\e[31mError: Don't create global factories/helpers"
  warn "Put #{crap_factories.map { |m| "`#{m}`" }.to_sentence} in the `Factories` module"
  warn "(or somewhere else appropriate)\e[0m"
  $stderr.puts
  exit! 1
end

File: ui/features/accessibility/accessibility_course_statistics/__tests__/factories.ts
/*
 * Copyright (C) 2026 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import type {Course, Teacher, Term, AccessibilityCourseStatistic} from '../types/course'

export const createMockTeacher = (overrides?: Partial<Teacher>): Teacher => ({
  id: '10',
  display_name: 'John Doe',
  html_url: '/users/10',
  avatar_image_url: 'https://example.com/avatar.jpg',
  ...overrides,
})

export const createMockTerm = (overrides?: Partial<Term>): Term => ({
  name: 'Fall 2026',
  ...overrides,
})

export const createMockAccessibilityCourseStatistic = (
  overrides?: Partial<AccessibilityCourseStatistic>,
): AccessibilityCourseStatistic => ({
  id: 1,
  course_id: 1,
  active_issue_count: 5,
  resolved_issue_count: 10,
  workflow_state: 'active',
  created_at: '2026-01-07T12:00:00Z',
  updated_at: '2026-01-07T12:00:00Z',
  ...overrides,
})

export const createMockCourse = (overrides?: Partial<Course>): Course => ({
  id: '1',
  name: 'Introduction to Testing',
  workflow_state: 'available',
  sis_course_id: 'SIS-101',
  total_students: 25,
  teachers: [
    createMockTeacher({id: '10', display_name: 'John Doe', html_url: '/users/10'}),
    createMockTeacher({id: '11', display_name: 'Jane Smith', html_url: '/users/11'}),
  ],
  term: createMockTerm(),
  subaccount_id: '5',
  subaccount_name: 'College of Engineering',
  accessibility_course_statistic: createMockAccessibilityCourseStatistic(),
  ...overrides,
})

export const createMockCourses = (count: number, overrides?: Partial<Course>): Course[] =>
  Array.from({length: count}, (_, i) =>
    createMockCourse({
      id: String(i),
      name: `Course ${i}`,
      ...overrides,
    }),
  )
